### PR TITLE
Optimized and Added New Features to the HoneyCombSolid Macro

### PR DESCRIPTION
**Pull request by eoyilmaz**
Thank you for creating a pull request to contribute to FreeCAD-macros!
To integrate your macro please make sure the following steps are complete:

- [X] Please check this box if you're not submitting a new macro.   
- [ ] Are you submitting a new macro ?  
- [x] Have you followed the ['How to submit a macro'](../README.md#how-to-submit-a-macro) section of the README.md ?  
- [x] Your macro has a [Description](../README.md#macro-description) in its header.  
- [x] Your macro has a [CamelCase name](../README.md#camelcase-macro-name).  
- [x] Your macro is named [appropriately](../README.md#macro-name-specifics).  
- [x] Your macro contains a [Metadata section](../README.md#macro-metadata) that immediately follows the header description.  
- [x] Your macro is Python3/Qt5 compliant and tested on the latest FreeCAD stable and development releases.  
- [ ] You're including documentation on how your macro works (bonus: screenshots and/or video on the Wiki)  
- [x] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)  
- [ ] Commit message is titled in the following way `[MacroName] Short description`.
- [ ] Optional, write or update the changelog in the macro, from latest to oldest.

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

Hi this is an update to the HoneyCombSolid Macro. I have done fixes to the general approach, added new features and optimized the code.

- Fixes:
  - For very big sizes the old code was not generating proper geometry.
    image
- New Features:
  - The old approach was doing an unnecessary boolean operation and it was not possible to only get the hexagonal shapes:
    image
  - The old one forces the user to generate an unnecessarily big hexagon grid if the user requires a custom outline. New approach allows the hexagon shapes to be combined with a custom shape.
  - This approach eliminates the unnecessary boolean operation and makes it work 4.5-5x faster.
  - The user can specify the usage of the container geometry with the ``UseContainer`` parameter.
  - The user can still switch back to the old algorithm by setting the ``UseNewAlgorithm`` parameter to False:
    image
- Optimizations:
  - As explained above not using a container is 4.5x to 5x faster:
    image
    image
   - And the new algorithm is 20% faster if a container is used: 
    image
  - The output matches the previous approach:
    image

### CODE REVIEW

File: ParametricObjectCreation/honeycomb_solid/honeycomb_solid.py [hunk 1]
```diff
@@ -163,7 +181,7 @@ def getDefaultDisplayMode(self):
         It must be defined in getDisplayModes.
 
         """
-        return 'Shaded'
+        return "Shaded"
```
Comment (galou) [inline ParametricObjectCreation/honeycomb_solid/honeycomb_solid.py:184]: Single-quotes are preferred in Python.

**Review by galou — APPROVED** [review]
Thanks for your contribution!

**Comment by christiangda** [comment]
> @christiangda Do you agree with the changes? They look good to me.

totally agree,  thanks

File: ParametricObjectCreation/honeycomb_solid/honeycomb_solid.py (hunk 1 above)
Comment (eoyilmaz) [reply ParametricObjectCreation/honeycomb_solid/honeycomb_solid.py:184]: Actually, I ran the code through ``black`` and it turned those single quotes two double quotes, as it is a very reputable linter that follows PEP8 I believe it is double quotes that is preferred. But, if this is against FreeCAD’s Python convention, I can revert them back to single quotes.

**Pull request merged**